packages/chain-ethereum-verifiable/src/EIP712VerifiableSigner.ts

```diff
@@ -0,0 +1,170 @@
+import { AbstractSigner, Wallet, hexlify, getBytes, verifyTypedData } from "ethers"
+import { logger } from "@libp2p/logger"
+
+import type { Signature, SessionSigner, Action, Message, Session } from "@canvas-js/interfaces"
+import { Secp256k1Signer, didKeyPattern } from "@canvas-js/signed-cid"
+
+import target from "#target"
+
+import {
+	eip712TypeDefinitionsForSession,
+	type EIP712VerifiableSessionData,
+	type EIP712VerifiableSessionMessage,
+} from "./types.js"
+import { assert, signalInvalidType, validateSessionData, parseAddress, addressPattern, generateSalt } from "./utils.js"
+
+export interface EIP712VerifiableSignerInit {
+	signer?: AbstractSigner
+	sessionDuration?: number
+	chainId?: number // used in the eip712 domain, but optional; no chainid if none is specified (don't default to mainnet)
+	verifyingContract?: string // used in the eip712 domain
+	salt?: string // used in the eip712 domain separator
+	version?: string // version in the eip712 domain. by default 1, but later versions of this signer could increment it
+}
+
+export class EIP712VerifiableSigner implements SessionSigner<EIP712VerifiableSessionData> {
+	public readonly sessionDuration: number | null
+	public readonly chainId: number
+	public readonly verifyingContract: string
+	public readonly salt: string
+	public readonly version: string
+
+	private readonly log = logger("canvas:chain-ethereum-verifiable")
+
+	#store = target.getSessionStore()
+	#ethersSigner: AbstractSigner
+
+	public constructor(init: EIP712VerifiableSignerInit = {}) {
+		this.#ethersSigner = init.signer ?? Wallet.createRandom()
+		this.sessionDuration = init.sessionDuration ?? null
+		this.chainId = init.chainId ?? 1
```

Comment:
for chainId and verifyContract, shouldn't the default be to leave them out? they are optional in the spec https://github.com/ethereum/EIPs/blob/master/EIPS/eip-712.md#definition-of-domainseparator

Reply:
yes that's right

Reply:
ok so re chainId, if it is not provided then how do we format the ethereum address? would it be something like `eip155::0x212421214...` or `eip155:null:0x212421214...` etc?